### PR TITLE
Mike/gate 1848 terraform support teams rules accounts

### PR DESCRIPTION
**Pull request by Michael9127**
Adds support for Cloudflare Teams Gateway account and rule configuration.

### CODE REVIEW

File: cloudflare/resource_cloudflare_teams_accounts_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "account_id", accountID),
+					resource.TestCheckResourceAttr(name, "tls_decrypt_enabled", "true"),
+					resource.TestCheckResourceAttr(name, "activity_log_enabled", "true"),
+					resource.TestCheckResourceAttr(name, "block_page.0.name", rnd),
+					resource.TestCheckResourceAttr(name, "block_page.0.enabled", "true"),
+					resource.TestCheckResourceAttr(name, "block_page.0.footer_text", "hello"),
+					resource.TestCheckResourceAttr(name, "block_page.0.header_text", "hello"),
+					resource.TestCheckResourceAttr(name, "block_page.0.background_color", "#000000"),
+					resource.TestCheckResourceAttr(name, "block_page.0.logo_path", "https://google.com"),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts_test.go:42]:
```suggestion
					resource.TestCheckResourceAttr(name, "block_page.0.logo_path", "https://example.com"),
```

File: cloudflare/resource_cloudflare_teams_accounts_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+resource "cloudflare_teams_account" "%[1]s" {
+  account_id  = "%[2]s"
+  tls_decrypt_enabled = true
+  activity_log_enabled = true
+  block_page {
+	name="%[1]s"
+	enabled=true
+    footer_text="hello"
+    header_text="hello"
+    logo_path="https://google.com"
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts_test.go:60]:
```suggestion
    logo_path="https://example.com"
```

File: cloudflare/resource_cloudflare_teams_rules.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+	"ytrestricted",
+	"on",
+	"off",
+	"scan",
+	"noscan",
+	"isolate",
+	"noisolate",
+	"override",
+	"l4_override",
+}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_rules.go:27]: this should probably live in `cloudflare-go` similar to Rulesets with a helper method to expose them as a string list.

- https://github.com/cloudflare/cloudflare-go/blob/master/rulesets.go#L20-L30
- https://github.com/cloudflare/cloudflare-go/blob/master/rulesets.go#L69-L83

File: cloudflare/resource_cloudflare_teams_rules.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+func resourceCloudflareTeamsRuleConfiguration() *schema.Resource {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_rules.go:29]:
```suggestion
func resourceCloudflareTeamsRule() *schema.Resource {
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/pkg/errors"
+)
+
+func resourceCloudflareTeamsAccountConfiguration() *schema.Resource {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:14]:
```suggestion
func resourceCloudflareTeamsAccount() *schema.Resource {
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+	return &schema.Resource{
+		Read:   resourceCloudflareTeamsAccountConfigurationRead,
+		Update: resourceCloudflareTeamsAccountConfigurationUpdate,
+		Create: resourceCloudflareTeamsAccountConfigurationUpdate,
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:18]:
```suggestion
		Read:   resourceCloudflareTeamsAccountRead,
		Update: resourceCloudflareTeamsAccountUpdate,
		Create: resourceCloudflareTeamsAccountUpdate,
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+		Type:     schema.TypeBool,
+		Required: true,
+	},
+	"fail_closed": {
+		Type:     schema.TypeBool,
+		Required: true,
+	},
+}
+
+func resourceCloudflareTeamsAccountConfigurationRead(d *schema.ResourceData, meta interface{}) error {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:100]:
```suggestion
func resourceCloudflareTeamsAccountRead(d *schema.ResourceData, meta interface{}) error {
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+	if configuration.Settings.ActivityLog != nil {
+		if err := d.Set("activity_log_enabled", configuration.Settings.ActivityLog.Enabled); err != nil {
+			return errors.Wrap(err, "error parsing account activity log enablement")
+		}
+	}
+	return nil
+}
+
+func resourceCloudflareTeamsAccountConfigurationUpdate(d *schema.ResourceData, meta interface{}) error {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:140]:
```suggestion
func resourceCloudflareTeamsAccountUpdate(d *schema.ResourceData, meta interface{}) error {
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+		updatedTeamsAccount.Settings.ActivityLog = &cloudflare.TeamsActivityLog{Enabled: activtyLog.(bool)}
+	}
+	log.Printf("[DEBUG] Updating Cloudflare Teams Account configuration from struct: %+v", updatedTeamsAccount)
+
+	if _, err := client.TeamsAccountUpdateConfiguration(context.Background(), accountID, updatedTeamsAccount); err != nil {
+		return fmt.Errorf("error updating Teams Account configuration for account %q: %s", accountID, err)
+	}
+
+	d.SetId(accountID)
+	return resourceCloudflareTeamsAccountConfigurationRead(d, meta)
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:168]:
```suggestion
	return resourceCloudflareTeamsAccountRead(d, meta)
```

File: cloudflare/resource_cloudflare_teams_accounts.go [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+	if len(attributes) != 2 {
+		return nil, fmt.Errorf("invalid id (\"%s\") specified, should be in format \"accountID/teamsAccountID\"", d.Id())
+	}
+
+	accountID := attributes[0]
+	log.Printf("[DEBUG] Importing Cloudflare Teams Account configuration for account %s", accountID)
+
+	d.Set("account_id", accountID)
+	err := resourceCloudflareTeamsAccountConfigurationRead(d, meta)
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts.go:182]:
```suggestion
	err := resourceCloudflareTeamsAccountRead(d, meta)
```

File: cloudflare/resource_cloudflare_teams_accounts_test.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+func TestAccCloudflareTeamsAccountConfigurationBasic(t *testing.T) {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_accounts_test.go:11]:
```suggestion
func TestAccCloudflareTeamsAccountBasic(t *testing.T) {
```

File: cloudflare/resource_cloudflare_teams_rules.go [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+	rule, err := client.TeamsRule(context.Background(), accountID, d.Id())
+	if err != nil {
+		if strings.Contains(err.Error(), "HTTP status 400") {
+			log.Printf("[INFO] Teams Rule config %s doesnt exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error finding Teams Rule %q: %s", d.Id(), err)
+	}
+	if err := d.Set("name", rule.Name); err != nil {
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_teams_rules.go:165]: nice error handling on all these fields. basic field types (strings, numbers, booleans) don't really need it but it's good to have either way. the real win is the complex types and if we have malformed data going into it.

**Review by jacobbednarz — CHANGES_REQUESTED** [review]
some small nitpicks on method naming for consistency but other than that, looks awesome!

File: cloudflare/resource_cloudflare_teams_accounts_test.go (hunk 11 above)
Comment (jacobbednarz) [reply cloudflare/resource_cloudflare_teams_accounts_test.go:11]: this test also seems to be failing the integration tests

```
TF_ACC=1 go test $(go list ./...) -v -run "^(TestAccCloudflareTeamsRuleBasic|TestAccCloudflareTeamsAccountConfigurationBasic)" -count 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareTeamsAccountConfigurationBasic
    testing.go:705: Step 0 error: Check failed: Check 3/9 error: cloudflare_teams_account.iaxdohpgah: Attribute 'activity_log_enabled' expected "true", got "false"
--- FAIL: TestAccCloudflareTeamsAccountConfigurationBasic (1.27s)
=== RUN   TestAccCloudflareTeamsRuleBasic
--- PASS: TestAccCloudflareTeamsRuleBasic (2.98s)
FAIL
FAIL	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	4.604s
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
FAIL
make: *** [testacc] Error 1
```

Comment (Michael9127) [reply cloudflare/resource_cloudflare_teams_accounts_test.go:11]: ahh that field is controlled by a flipper that is enabled on my account. I removed it from the test.

File: cloudflare/resource_cloudflare_teams_rules.go (hunk 3 above)
Comment (Michael9127) [reply cloudflare/resource_cloudflare_teams_rules.go:27]: This makes sense, but would require waiting for another cloudflare-go release next month. Can we merge it for now as-is and ill be sure to make a ticket and put up a PR to take care of it later?

Comment (jacobbednarz) [reply cloudflare/resource_cloudflare_teams_rules.go:27]: sure, sounds like a plan 👍

**Comment by jacobbednarz** [comment]
just needing the registry documentation here and we can merge! see a good example in #1113

**Comment by Michael9127** [comment]
Updated documentation for this PR and my previous one for Teams Locations where I forgot to.

File: cloudflare/resource_cloudflare_teams_rules.go (hunk 3 above)
Comment (Michael9127) [reply cloudflare/resource_cloudflare_teams_rules.go:27]: https://github.com/cloudflare/cloudflare-go/pull/702

**Comment by jacobbednarz** [comment]
i pushed up some fixes for the documentation and `Import` functionality but this looks great! integration is all green too.

```
TF_ACC=1 go test $(go list ./...) -v -run "^(TestAccCloudflareTeamsAccountConfigurationBasic|TestAccCloudflareTeamsRuleBasic)" -count 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareTeamsAccountConfigurationBasic
--- PASS: TestAccCloudflareTeamsAccountConfigurationBasic (1.75s)
=== RUN   TestAccCloudflareTeamsRuleBasic
--- PASS: TestAccCloudflareTeamsRuleBasic (2.79s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	5.192s
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

**Pull request merged**